src/circuit/hop.js

```diff
@@ -106,11 +107,12 @@ class Hop extends EE {
     series([
       (cb) => this.utils.validateAddrs(message, sh, proto.Type.HOP, cb),
       (cb) => isConnected(cb),
-      (cb) => this._circuit(sh.rest(), message, cb)
+      (cb) => this._circuit(sh, message, cb)
     ], (err) => {
       if (err) {
         log.err(err)
-        setImmediate(() => this.emit('circuit:error', err))
+        sh.close()
```

Comment:
I ran the libp2p, switch and interop tests for this update and switch was the only thing that found an issue.

This test https://github.com/libp2p/js-libp2p-switch/blob/v0.40.8/test/circuit-relay.node.js#L322-L329 had the issue.

What's happening is when the circuit dials to the peer, it gets the muxer error for not supporting relay. This causes the stream to be closed at https://github.com/libp2p/js-libp2p-circuit/blob/27593926fa221d8c520d9dfd227fb47d31e82f01/src/circuit/hop.js#L213. That error is passed to L112, where the stream is closed again. Since it was already closed, there is no longer a shake in the stream handler, https://github.com/libp2p/js-libp2p-circuit/blob/27593926fa221d8c520d9dfd227fb47d31e82f01/src/circuit/stream-handler.js#L111, so `.rest` throws an error:

```
     Uncaught TypeError: Cannot read property 'rest' of null
      at StreamHandler.rest (node_modules/libp2p-circuit/src/circuit/stream-handler.js:111:29)
      at StreamHandler.close (node_modules/libp2p-circuit/src/circuit/stream-handler.js:128:12)
```

We should add a test to this repo to replicate the issue so we catch future issues.

Reply:
Thanks for catching this. I'll add a fix and the test case 👍

Reply:
@jacobheun I've pushed the fixes for the `SH.close()` issues - it now allows to be called more than once without crashing. 

I've also added missing cb returns in stream handler, that should prevent it from going into a bad state. This should address the issues reported in:

- https://github.com/libp2p/js-libp2p-circuit/issues/36
- https://github.com/libp2p/js-libp2p-circuit/issues/35
- https://github.com/libp2p/js-libp2p-circuit/issues/34

However, I realize that whats needed are stream-handler tests, which are currently missing, I'd love to add them in, but I'm a bit low on time right now. I'll open up an issue to keep track of that.